src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java

```diff
@@ -18,7 +26,231 @@
     }
+
+    public RootTransformBuilder rootTransform()
+    {
+        return rootTransform;
+    }
+
+    @Override
+    public JsonObject toJson()
+    {
+        JsonObject json = super.toJson();
+
+        // Write the transform to an object
+        JsonObject transform = new JsonObject();
+
+        if (!rootTransform.translation.equals(Vector3f.ZERO))
+        {
+            transform.add("translation", writeVec3(rootTransform.translation));
+        }
+
+        if (!rootTransform.scale.equals(new Vector3f(1, 1, 1)))
+        {
+            transform.add("scale", writeVec3(rootTransform.scale));
+        }
+
+        Quaternion leftRotation = rootTransform.leftRotation;
+        if (!leftRotation.equals(Quaternion.ONE))
+        {
+            transform.add("left_rotation", writeQuaternion(leftRotation));
+        }
+
+        Quaternion rightRotation = rootTransform.rightRotation;
+        if (!rightRotation.equals(Quaternion.ONE))
+        {
+            transform.add("right_rotation", writeQuaternion(rightRotation));
+        }
+
+        if (rootTransform.originName != null)
+        {
+            transform.addProperty("origin", rootTransform.originName);
+        }
+        else if (rootTransform.origin != null && rootTransform.origin != Vector3f.ZERO)
+        {
+            transform.add("origin", writeVec3(rootTransform.origin));
+        }
+
+        // If there were any transform properties set, add them to the output.
+        if (transform.size() > 0)
+        {
+            json.add("transform", transform);
+        }
+
+        return json;
+    }
+
+    private JsonArray writeVec3(Vector3f vector)
+    {
+        JsonArray array = new JsonArray();
+        array.add(vector.x());
+        array.add(vector.y());
+        array.add(vector.z());
+        return array;
+    }
+
+    private JsonArray writeQuaternion(Quaternion quaternion)
+    {
+        JsonArray array = new JsonArray();
+        array.add(quaternion.i());
+        array.add(quaternion.j());
+        array.add(quaternion.k());
+        array.add(quaternion.r());
+        return array;
+    }
+
+    public class RootTransformBuilder
+    {
+        private Vector3f translation = Vector3f.ZERO;
+        private Quaternion leftRotation = Quaternion.ONE.copy();
+        private Quaternion rightRotation = Quaternion.ONE.copy();
+        private Vector3f scale = new Vector3f(1, 1, 1);
+
+        private @Nullable String originName;
+        private @Nullable Vector3f origin;
+
+        /**
+         * Sets the translation of the root transform.
+         *
+         * @param translation the translation
+         * @return this builder
+         * @throws NullPointerException if {@code translation} is {@code null}
+         */
+        public RootTransformBuilder translation(Vector3f translation)
+        {
+            Preconditions.checkNotNull(translation, "Translation must not be null");
+            this.translation = translation;
```

Comment:
In general for all these methods, the `checkNotNull` can be inlined to the field assignment as it returns the value passed when not `null`.